rest/src/main/java/com/epam/esm/rest/CertificateController.java

```diff
@@ -0,0 +1,161 @@
+import com.epam.esm.service.exception.ServiceValidationException;
+import com.epam.esm.transferobj.FilterParam;
+import com.epam.esm.transferobj.OrderParam;
+import com.epam.esm.transferobj.ParameterConstant;
+
+@RestController
+@RequestMapping("/certificates")
+public class CertificateController {
+
+	private CertificateService certificateService;
+	private MessageSource messageSource;
+	private static final Logger log = LogManager.getLogger(CertificateController.class);
+
+	@Autowired
+	public CertificateController(CertificateService certificateService, MessageSource messageSource) {
+		this.certificateService = certificateService;
+		this.messageSource = messageSource;
+	}
+
+	/**
+	 * GET certificate by the long Id;
+	 * 
+	 * @param certificateId
+	 * @return {@link GiftCertificateGetDTO} (in case when the certificate with the
+	 *         given Id is not found, the method returns Status Code = 404)
+	 */
+	@GetMapping("{certificateId}")
+	public GiftCertificateGetDTO getSertificate(@PathVariable long certificateId) {
+		GiftCertificateGetDTO giftCertificate = certificateService.getCertificate(certificateId);
+		if (giftCertificate == null) {
+			throw new NotFoundException(messageSource.getMessage((MessageKeyHolder.CERTIFICATE_NOT_FOUND_KEY),
+					new Object[] { certificateId }, Locale.getDefault()));
+		}
+		return giftCertificate;
+	}
+
+	/**
+	 * POST method which adds new certificate into the Database;
+	 * 
+	 * @param certificate The argument should contains the List of tags to bound the
+	 *                    certificate with
+	 * @return {@link GiftCertificateGetDTO} (in case of success, the method returns
+	 *         Status Code = 200 and the response body contains new generated
+	 *         certificates's Id)
+	 */
+	@PostMapping
+	public GiftCertificateGetDTO addCertificate(@Valid @RequestBody GiftCertificateCreateUpdateDTO certificate) {
+		if (certificate.getTags() == null || certificate.getTags().isEmpty()) {
+			throw new InvalidRequestParametersException(messageSource
+					.getMessage((MessageKeyHolder.CERTIFICATE_INVALID_TAGS_KEY), null, Locale.getDefault()));
+		}
+		return certificateService.saveCertificate(certificate);
+	}
+
+	/**
+	 * PUT method which allows to change the certificate's Name, Description, Price,
+	 * Duration and the tags to bound with;
+	 * 
+	 * @param certificateId, certificate
+	 * @return {@link GiftCertificateGetDTO} (in case of success, the method returns
+	 *         Status Code = 200)
+	 */
+	@PutMapping("{certificateId}")
+	public GiftCertificateGetDTO updateCertificate(@PathVariable long certificateId,
+			@Valid @RequestBody GiftCertificateCreateUpdateDTO certificate) {
+		GiftCertificateGetDTO giftCertificateGetDTO = certificateService.getCertificate(certificateId);
+		if (giftCertificateGetDTO == null) {
+			throw new NotFoundException(messageSource.getMessage((MessageKeyHolder.CERTIFICATE_NOT_UPDATED_KEY),
+					new Object[] { certificateId }, Locale.getDefault()));
```

Comment:
It looks like a validation. If it is, I believe it is a responsibility of the server layer.